### ISSUE TITLE
Remove unused automatically added using directives on save

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#66898]
## Summary
 
Add a new default behavior to automatically remove any unused using directives on save which were implicitly added to the document by another feature (e.g. add usings on paste, completion for unimported types).

## Background and Motivation

Sometimes users unintentionally complete items that result in using directives being added, or change course while developing such that the automatically-added using directive is no longer needed. There is no mechanism is place to automatically clean these up.

## Proposed Feature

The new feature would keep track of implicitly-added using directives for an unsaved document. When the document is saved, any of these directives which are no longer in use will be automatically removed.

Manually added using directives will not be altered by this feature, even in cases where they are not used. This includes indirectly added using directives from features like the Add Using Directive code fix.

## Alternative Designs

Code Cleanup on Save is intended to fulfill this need, but is too aggressive for many users and too slow for others.

**Comment by Dean-NC** [comment]
Don't forget that there's a bug where if you paste code that references a type that's not defined in the project, instead of not putting any using for that unknown type (correct), it puts a using for `System.Numerics` or other random one (incorrect). So in this case, instead of removing the wrong one after the fact, just don't put it in.

**Comment by CyrusNajmabadi** [comment]
@Dean-NC can you give an example (or report a new bug that demonstrates the issue)?

**Comment by sharwell** [comment]
@Dean-NC That's a very different bug. I'll see if I can find a reference.

**Comment by Dean-NC** [comment]
@CyrusNajmabadi @sharwell  I've been providing details about the paste bug in the following VS bug report that was closed as being a duplicate of this Roslyn issue.
[https://developercommunity.visualstudio.com/t/Visual-Studio-2022-keeps-adding-unnecess/10226615](https://developercommunity.visualstudio.com/t/Visual-Studio-2022-keeps-adding-unnecess/10226615)
I didn't create that report, but I provided most of the detail, including the animated .Gif toward the end that shows the bug I've mentioned. I didn't know when I used that bug report that there were different bugs. It was the only report about erroneous usings I could find at the time. I tried to let the support person in that report know what I found, but they went **dark**.

You were quick to mark this as off-topic.
Thanks for the correct Roslyn issue link.

**Comment by smaillet-ms** [comment]
It would be best if this was implemented with an option to disable ALL "implicit" auto insertion of usings. They should only be added when I explicitly want them to be. (And I can select the one I want if there are ambiguities). I don't want the IDE ever trying to guess my intent - I know it better (No matter how much AI you put in to things, you can't read my mind) I get a TON of these all the time and it is REALLY annoying to have to constantly remove things. (I just got this one "
using static System.Runtime.InteropServices.JavaScript.JSType;" But my code has NOTHING at all to do with that, it's not even referenced anywhere.  I have no idea what that is or where things like that keep coming from. I have as much auto insert xyz turned off as I possibly can. IntelliSense doesn't pop up unless I ask it to. So, I'm confused by claims this is a result of "completing" something unintentionally.

**Comment by CyrusNajmabadi** [comment]
@smaillet-ms  You can turn this off with:

![image](https://github.com/dotnet/roslyn/assets/4564579/a2a6aec9-2bda-4f9a-bb04-1ff9c6851999)

and

![image](https://github.com/dotnet/roslyn/assets/4564579/f6bc0023-2680-4a76-8167-d63d6085696c)

**Comment by CyrusNajmabadi** [comment]
> Oh one more thing, these source files are not even opened in VS when the references are being added.

@sej69 This is not very unlikely to be roslyn.  Perhaps some other tool or extension you're using.

**Comment by RobDaytona** [comment]
Just to add my comments here, Maui is doing this in my new project aslo. I have been manaully removing these weird using statements that seem to randomly appear when compiling or after compiling or who knows when exactly..
for example this one just popped up:
using Intents;

**Comment by RobDaytona** [comment]
today it randomly added this to into a random class's usings:
using MetalPerformanceShaders;

**Comment by CyrusNajmabadi** [comment]
@RobDaytona see my message above for how to disable these features if you do not want them.

**Comment by RobDaytona** [comment]
Thanks for that @CyrusNajmabadi

**Comment by clovett** [comment]
This is impacting my productivity, I recently found it inserted "using Azure.AI.OpenAI" when I never used any such thing.  I'm also finding the intellisense is fighting me more than normal, probably trying to be too clever.  I think I'll try running with "Show items from unimported namespaces" turn off for a while, that might be the issue.

**Comment by CyrusNajmabadi** [comment]
> I'm also finding the intellisense is fighting me more than normal

If you run into issues, please file then @clovett 

>  I recently found it inserted "using Azure.AI.OpenAI" when I never used any such thing.

This generally means you explicitly committed a completion item suggestion for a type from that namespace.

**Comment by clovett** [comment]
> This generally means you explicitly committed a completion item suggestion for a type from that namespace.

Right but I'm finding that as I type really fast the whole busy prompting thing gets in the way so I can't type as fast as I used to since I have to spend all my time reading useless prompts to know when to not hit TAB.

**Comment by CyrusNajmabadi** [comment]
I don't know what you mean by "busy prompting" :-)

> to know when to not hit TAB.

The only new feature we've added here in Roslyn is the suggestion of types from other namespaces (one of our largest customer requests).  But that feature can be turned off, returning to the same intellisense experience for the last 20ish years :-)

**Comment by clovett** [comment]
Yes, I've been running with this turned off most of today and things have been better.  No more random using and unwanted tab completions.  It feels like there must be something other than TAB that is completing things, like "(" or "," or ";".  So I'm typing fast, it injects something unwanted while I'm typing what I know to be correct and before I know it when I type "(" to open a function call, it has completed something else and injected the unwanted using statement.  In this  "using Azure.AI.OpenAI"  case it also crept in a dependency that ended up getting packaged in my nuget package, woops!!

**Comment by pj6** [comment]
The problem statement seems to be worded incorrectly: pasting short snippets of valid code often generates completely unrelated using statements. This is not the user "unintentionally completing items" or "changing course". It's incorrect behavior. This feature should not be generating incorrect usings in the first place.

If we can't change the wording of the problem here, or the proposed solution, you should allow us to break requests for _real_ fix out into a separate issue instead of rolling it into this one. Because this is not a fix, it's a band-aid.

**Comment by CyrusNajmabadi** [comment]
@pj6 Def file bugs if you're running into issues.  I looked in our issue tracker and i don't see anything reported by you.  So i'm not sure what you're referring to here.

**Comment by pj6** [comment]
@CyrusNajmabadi It will just get rolled into this one re: the second paragraph of my comment.

The "paste bug"
[https://developercommunity.visualstudio.com/t/Visual-Studio-2022-keeps-adding-unnecess/10226615](https://developercommunity.visualstudio.com/t/Visual-Studio-2022-keeps-adding-unnecess/10226615)

Even without that example, I have _repeatedly_ had bug reports inappropriately rolled into other issues (not so much with this account, but my work one). And there's no recourse, no consideration to disagreement after the decision's been made. I guarantee you if I enter this as a separate bug it will be duped to here and they just won't listen at all.

Which is why this discussion belongs _here_, until we fix _that_ issue too.

**Comment by CyrusNajmabadi** [comment]
Please file an issue on Roslyn with a repro. Thanks.

**Comment by CyrusNajmabadi** [comment]
> Which is why this discussion belongs here, until we fix that issue too.

This discussion is about a different topic. Please open an issue against Roslyn for the bug you're seeing. Thanks.
